src/coreclr/jit/promotion.cpp

```diff
@@ -675,6 +674,505 @@ class LocalsUseVisitor : public GenTreeVisitor<LocalsUseVisitor>
+                // no longer dying here.
+                addr->gtFlags &= ~GTF_VAR_DEATH;
+            }
+            else if (addr->IsInvariant())
+            {
+                // Fall through
+            }
+            else
+            {
+                unsigned addrLcl = m_compiler->lvaGrabTemp(true DEBUGARG("Spilling address for field-by-field copy"));
+                statements->AddStatement(m_compiler->gtNewTempAssign(addrLcl, addr));
+                addr = m_compiler->gtNewLclvNode(addrLcl, addr->TypeGet());
+                UpdateEarlyRefCount(m_compiler, addr);
+            }
+        }
+
+        auto grabAddr = [&numAddrUses, addr, this](unsigned offs) {
+            assert(numAddrUses > 0);
+            numAddrUses--;
+
+            GenTree* addrUse;
+            if (numAddrUses == 0)
+            {
+                // Last use of the address, reuse the node.
+                addrUse = addr;
+            }
+            else
+            {
+                addrUse = m_compiler->gtCloneExpr(addr);
+                UpdateEarlyRefCount(m_compiler, addrUse);
+            }
+
+            if (offs != 0)
+            {
+                var_types addrType = varTypeIsGC(addrUse) ? TYP_BYREF : TYP_I_IMPL;
+                addrUse            = m_compiler->gtNewOperNode(GT_ADD, addrType, addrUse,
+                                                    m_compiler->gtNewIconNode((ssize_t)offs, TYP_I_IMPL));
+            }
+
+            return addrUse;
+        };
+
+        if (!coversDestination)
+        {
+            // Note that this does not handle partially overlapping copies,
+            // but that is left undefined (and normal block copies do not
+            // handle this either).
+            if (m_src->OperIs(GT_BLK, GT_FIELD))
+            {
+                // Note that we should use 0 instead of addrBaseOffs here
+                // since this ends up as the address of the GT_FIELD node
+                // that already has the field offset.
+                m_src->AsUnOp()->gtOp1 = grabAddr(0);
+            }
+            else if (m_dst->OperIs(GT_BLK, GT_FIELD))
+            {
+                // Like above, use 0 intentionally here.
+                m_dst->AsUnOp()->gtOp1 = grabAddr(0);
+            }
+
+            statements->AddStatement(m_compiler->gtNewBlkOpNode(m_dst, m_src));
+        }
+
+        if (needsNullCheck)
+        {
+            statements->AddStatement(m_compiler->gtNewIndir(TYP_BYTE, grabAddr(addrBaseOffs)));
+        }
+
+        for (int i = 0; i < m_entries.Height(); i++)
+        {
+            const Entry& entry = m_entries.BottomRef(i);
+
+            GenTree* dst;
+            if (entry.ToLclNum != BAD_VAR_NUM)
+            {
+                dst = m_compiler->gtNewLclvNode(entry.ToLclNum, entry.Type);
+
+                if (m_compiler->lvaGetDesc(entry.ToLclNum)->lvIsStructField)
+                    UpdateEarlyRefCount(m_compiler, dst);
```

Comment:
This looks a bit odd right now because I'm trying to avoid adding any ref counts to the replacement locals created by this pass. So this logic tracks them correctly only for existing locals and relies on forward sub to give up on the physical promotion locals that have 0 ref count.

I will submit a separate PR after this one to correctly track the ref counts much more uniformly for all the IR created by this pass.